pkg/tsdb/cloudmonitoring/mql_query.go

```diff
@@ -0,0 +1,351 @@
+package cloudmonitoring
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/grafana/grafana/pkg/components/null"
+	"github.com/grafana/grafana/pkg/components/simplejson"
+	"github.com/grafana/grafana/pkg/tsdb"
+	"github.com/grafana/grafana/pkg/tsdb/sqleng"
+	"github.com/opentracing/opentracing-go"
+	"golang.org/x/net/context/ctxhttp"
+)
+
+func (query cloudMonitoringMqlQuery) executeQuery(ctx context.Context, tsdbQuery *tsdb.TsdbQuery, e *CloudMonitoringExecutor) (*tsdb.QueryResult, cloudMonitoringResponse, error) {
+	queryResult := &tsdb.QueryResult{Meta: simplejson.New(), RefId: query.RefID}
+	projectName := query.ProjectName
+	if projectName == "" {
+		defaultProject, err := e.getDefaultProject(ctx)
+		if err != nil {
+			queryResult.Error = err
+			return queryResult, cloudMonitoringResponse{}, nil
+		}
+		projectName = defaultProject
+		slog.Info("No project name set on query, using project name from datasource", "projectName", projectName)
+	}
+
+	from, err := tsdbQuery.TimeRange.ParseFrom()
+	if err != nil {
+		queryResult.Error = err
+		return queryResult, cloudMonitoringResponse{}, nil
+	}
+	to, err := tsdbQuery.TimeRange.ParseTo()
+	if err != nil {
+		queryResult.Error = err
+		return queryResult, cloudMonitoringResponse{}, nil
+	}
+	intervalCalculator := tsdb.NewIntervalCalculator(&tsdb.IntervalOptions{})
+	interval := intervalCalculator.Calculate(tsdbQuery.TimeRange, time.Duration(query.IntervalMs/1000)*time.Second)
+	timeFormat := "2006/01/02-15:04:05"
+	query.Query += fmt.Sprintf(" | graph_period %s | within d'%s', d'%s'", interval.Text, from.UTC().Format(timeFormat), to.UTC().Format(timeFormat))
+
+	buf, err := json.Marshal(map[string]interface{}{
+		"query": query.Query,
+	})
+	if err != nil {
+		queryResult.Error = err
+		return queryResult, cloudMonitoringResponse{}, nil
+	}
+	req, err := e.createRequest(ctx, e.dsInfo, fmt.Sprintf("cloudmonitoring%s", "v3/projects/"+projectName+"/timeSeries:query"), bytes.NewBuffer(buf))
```

Comment:
The following should be a lot cleaner:
```suggestion
	req, err := e.createRequest(ctx, e.dsInfo, path.Join("cloudmonitoringv3/projects", projectName, "timeSeries:query"), bytes.NewBuffer(buf))
```